subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <stdint.h>
+#include <zephyr.h>
+#include <flash.h>
+#include <download_client.h>
+#include <nrf_socket.h>
+#include <dfu/mcuboot.h>
+#include <pm_config.h>
+#include <bsd.h>
+#include <logging/log.h>
+#include <fota_download.h>
+
+LOG_MODULE_REGISTER(fota_download, CONFIG_FOTA_DOWNLOAD_LOG_LEVEL);
+
+static struct	device *flash_dev;
+static bool	is_flash_page_erased[FLASH_PAGE_MAX_CNT];
+static		fota_download_callback_t callback;
+static struct	download_client dfu;
+
+static void report_flash_err(int err)
+{
+	struct fota_download_evt evt = {.id = FOTA_DOWNLOAD_FLASH_ERR,
+					.flash_error_code = err};
+	callback(&evt);
+}
+
+static void report_dlc_evt(const struct download_client_evt *dlc_evt)
+{
+	struct fota_download_evt evt = {.id = FOTA_DOWNLOAD_EVT_DOWNLOAD_CLIENT,
+					.dlc_evt = dlc_evt};
+	callback(&evt);
+}
+
+static int flash_page_erase_if_needed(u32_t address)
+{
+	int err;
+	struct flash_pages_info info;
+
+	err = flash_get_page_info_by_offs(flash_dev, address, &info);
+	if (err != 0) {
+		LOG_ERR("flash_get_page_info_by_offs returned error %d\n", err);
+		return err;
+	}
+	if (!is_flash_page_erased[info.index]) {
+		err = flash_write_protection_set(flash_dev, false);
+		if (err != 0) {
+			LOG_ERR("flash_write_protection_set error %d\n", err);
+			return err;
+		}
+		err = flash_erase(flash_dev, info.start_offset, info.size);
+		if (err != 0) {
+			LOG_ERR("flash_erase error %d at address %08x\n",
+				err, info.start_offset);
+			return err;
+		}
+		is_flash_page_erased[info.index] = true;
+		err = flash_write_protection_set(flash_dev, true);
+		if (err != 0) {
+			LOG_ERR("flash_write_protection_set error %d\n", err);
+			return err;
+		}
+	}
+	return 0;
+}
+
+static int download_client_callback(const struct download_client_evt *event)
+{
+	int err;
+	static u32_t flash_address = PM_MCUBOOT_SECONDARY_ADDRESS;
+
+	__ASSERT(event != NULL, "invalid dfu object\n");
+
+	switch (event->id) {
+	case DOWNLOAD_CLIENT_EVT_FRAGMENT: {
+		size_t size;
+
+		err = download_client_file_size_get(&dfu, &size);
+		if (err != 0) {
+			LOG_ERR("download_client_file_size_get returned error"
+				"%d\n", err);
+			return err;
+		}
+		if (size > PM_MCUBOOT_SECONDARY_SIZE) {
+			LOG_ERR("Requested file too big to fit in flash\n");
+			report_flash_err(-EFBIG);
+			return -EFBIG;
+		}
+
+		err = flash_page_erase_if_needed(flash_address);
+		if (err != 0) {
+			report_flash_err(err);
+			return err;
+		}
+
+		err = flash_write_protection_set(flash_dev, false);
+		if (err != 0) {
+			LOG_ERR("flash_write_protection_set error %d\n", err);
+			report_flash_err(err);
+			return err;
+		}
+
+		err = flash_write(flash_dev, flash_address,
+				event->fragment.buf, event->fragment.len);
+		if (err != 0) {
+			LOG_ERR("Flash write error %d at address %08x\n",
+				err, flash_address);
+			report_flash_err(err);
```

Comment:
I think you should restore the write protection in case of failure here.

Reply:
Done, how should an error in this call be handled?

Reply:
That's actually a good question, on how to proceed with an error in error handling procedure :-) I guess we need flash expert to tell us if it's critical - if not I guess log is enough.

Reply:
@nvlsianpu  do you have any comments regarding this?

Reply:
NRF flash driver emulates write protection - API call has no reflection in peripheral register operation until flash_write/erase are called. So can even left it switched off.

a flash_write call might return error:
* on bad arguments
* if something is really not good with the system (for instance can not synchronize with a radio driver properly).
This shouldn't happened at all, but if happen (let say we have unstable synchronization :(, or caught cosmic ray particle) written flash content might be written partially. It is possible to check what is missing and retry to write missing payload (up to a few times). Although return with failure is the simplest way.

Reply:
I see, looks like we don't need to worry about checking this.

```
static int flash_nrf_write_protection(struct device *dev, bool enable)
{
	/* virtual write-erase protection */
	write_protect = enable;

	return 0;
}
```

Reply:
yeap